gossip/emitter/throttler/dominant_set_test.go

```diff
@@ -0,0 +1,245 @@
+	}{
+		"no validators": {
+			stakes: nil,
+		},
+		"single validator": {
+			stakes:      []int64{100},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"two equal validators": {
+			stakes:      []int64{50, 50},
+			expectedSet: []idx.ValidatorID{1, 2},
+		},
+		"two validators one dominant": {
+			stakes:      []int64{80, 20},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"three validators one dominant": {
+			stakes:      []int64{80, 10, 10},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"three validators two dominant": {
+			stakes:      []int64{40, 40, 20},
+			expectedSet: []idx.ValidatorID{1, 2},
+		},
+		"four equal validators, first three dominate": {
+			stakes:      []int64{25, 25, 25, 25},
+			expectedSet: []idx.ValidatorID{1, 2, 3},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			stakes := makeValidatorsFromStakes(test.stakes...)
+			set := ComputeDominantSet(stakes, pos.Weight(100), testThreshold)
+			require.ElementsMatch(t, test.expectedSet, slices.Collect(maps.Keys(set)))
+		})
+	}
+}
+
+func TestComputeDominantSet_UniformDistributionsAreDominated(t *testing.T) {
+	// A set of n validators with equal stake shall have a dominant set
+	// of size ceil(n * threshold)
+
+	for threshold := float64(0.67); threshold <= 1.; threshold += 0.05 {
+		for n := 1; n <= 100; n++ {
+			t.Run(fmt.Sprintf("n=%d,threshold=%.2f", n, threshold), func(t *testing.T) {
+				validators := makeValidatorsFromStakes(slices.Repeat([]int64{10}, n)...)
+				set := ComputeDominantSet(validators, validators.TotalWeight(), threshold)
+				expectedCount := int(math.Ceil(float64(n) * threshold))
+				require.Len(t, set, expectedCount, "n=%d, threshold=%.2f", n, threshold)
+				require.GreaterOrEqual(t, sumStake(set, validators), validators.TotalWeight()*pos.Weight(threshold))
+			})
+		}
+	}
+}
+
+func TestComputeDominantSet_CanBeEmpty(t *testing.T) {
+
+	t.Run(
+		"nominal stake > than online stake * threshold",
+		func(t *testing.T) {
+			oneValidator := makeValidatorsFromStakes(100)
+			for threshold := float64(0.67); threshold <= 1.; threshold += 0.05 {
+				nominalWeight := float64(oneValidator.TotalWeight())/threshold + 1
+				x := ComputeDominantSet(oneValidator, pos.Weight(nominalWeight), threshold)
+				require.Empty(t, x,
+					"if nominalStake * threshold > sum of validator stakes, the dominant set cannot exist")
+			}
+		})
+
+	t.Run("threshold > 1.0", func(t *testing.T) {
+		stakes := slices.Repeat([]int64{42}, 11)
+		validators := makeValidatorsFromStakes(stakes...)
+
+		threshold := 1.01
+		set := ComputeDominantSet(validators, validators.TotalWeight(), threshold)
+		require.Empty(t, set)
+	})
+}
+
+func TestComputeDominantSet_IsIndependentFromStakeOrder(t *testing.T) {
+	// The dominant set calculation does not sort validators by stake,
+	// this is done by the [pos.Validators] object itself. Nevertheless the code
+	// is highly dependent on this behavior, so we test it here.
```

Comment:
my first thought was "if it is independent from the stake order, should the expected set be always the same?" but then I realized that the order is based on validator id not stake. Should this be commented somewhere in `ComputeDominantSet`'s implementation? since it highly depends on this order ?